File: clojure/difference-of-squares/src/difference_of_squares.clj
(ns difference-of-squares)

(defn- square [n] (* n n))

(defn square-of-sums [n]
  (->> (range 1 (inc n))
       (reduce +)
       square))

(defn sum-of-squares [n]
  (->> (range 1 (inc n))
       (map square)
       (reduce +)))

(defn difference [n]
  (- (square-of-sums n) (sum-of-squares n)))

File: javascript/difference-of-squares/difference-of-squares.js
function Squares (num) {
    'use strict';

    function range (val) {
        var ret = new Array(val);
        for (var i = 0; i < val; i ++) {
            ret[i] = i + 1;
        }
        return ret;
    }

    function sumOfSquares (val) {
        console.log('call sumOfSquares');
        var r = range(val);
        return r.reduce(function (pval, cval) {
            return pval + Math.pow(cval, 2);
        }, 0);
    }

    function squareOfSums (val) {
        console.log('call squareOfSums');
        var r = range(val);
        var sum = r.reduce(function (pval, cval) {
            return pval + cval;
        }, 0);
        return Math.pow(sum, 2);
    }

    console.log('call constructor');
    return {
        squareOfSums: squareOfSums(num),
        sumOfSquares: sumOfSquares(num),
        difference: squareOfSums(num) - sumOfSquares(num)
    };
}

module.exports = Squares;

